### PR TITLE
[FLINK-12535][network] Make CheckpointBarrierHandler non-blocking

### PR DESCRIPTION
**Pull request by pnowojski**
This change makes `CheckpointBarrierHandler` a non blocking class with similar interface as `InputGate`. This is one step towards a mailbox model.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by pnowojski** [comment]
Test failure is unrelated and caused by: https://issues.apache.org/jira/browse/FLINK-12512

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputGate.java [hunk 1]
```diff
@@ -78,32 +79,32 @@
...
-
 	public abstract void requestPartitions() throws IOException, InterruptedException;
 
 	/**
 	 * Blocking call waiting for next {@link BufferOrEvent}.
 	 *
 	 * @return {@code Optional.empty()} if {@link #isFinished()} returns true.
 	 */
-	public abstract Optional<BufferOrEvent> getNextBufferOrEvent() throws IOException, InterruptedException;
+	public abstract Optional<BufferOrEvent> getNext() throws IOException, InterruptedException;
```
Comment (AjaxXu) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputGate.java]: Even through get a buffer or event without blocking, it still can be named with getNextBufferOrEvent(). What do you think?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ * Interface defining couple of essential methods for asynchronous and non blocking data polling.
+ */
+public interface AsyncDataInput<T> {
+	CompletableFuture<?> AVAILABLE = CompletableFuture.completedFuture(null);
+
+	/**
+	 * Poll the next element.
+	 *
+	 * @return If the {@link T} is an {@code Optional}:
+	 * 1. return {@code Optional.of(element)} - next returned element. {@code Optional.empty()} will
```
Comment (StefanRRichter) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java]: Feel like the 1. & 2. is on he wrong level which confused me a little. I would rather do 1. Optional ... 2. Otherwise and in that case the enumeration could also be dropped.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputGate.java [hunk 3]
```diff
@@ -68,7 +69,7 @@
  * will have an input gate attached to it. This will provide its input, which will consist of one
  * subpartition from each partition of the intermediate result.
  */
-public abstract class InputGate implements AutoCloseable {
+public abstract class InputGate implements AsyncDataInput<Optional<BufferOrEvent>>, AutoCloseable {
 
 	public static final CompletableFuture<?> AVAILABLE = CompletableFuture.completedFuture(null);
```
Comment (sunhaibotb) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputGate.java]: The `InputGate#AVAILABLE`  should be removed because the new `AsyncDataInput` interface already has `AVAILABLE`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	 * be returned if there is no data to return or if {@link #isFinished()} returns true.
+	 * otherwise:
+	 * 2. next element or null if there is no data to return or if {@link #isFinished()} returns true.
+	 */
+	T pollNext() throws Exception;
```
Comment (StefanRRichter) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java]: Minor suggestion: personal taste, I usually like to have such methods named like `tryNext()`, but I know that `poll` is also very common. You can adapt it if you find that slightly more clear as well.

Comment (StefanRRichter) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java]: Exception is not mentioned in the doc, even if it is rather clear what that is about.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Interface defining couple of essential methods for asynchronous and non blocking data polling.
```
Comment (StefanRRichter) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java]: I would suggest to outline the intended usage here in the class level doc:
Consume via poll until you get some representation of null or empty, check if the data is exhausted and if not wait on the future. This might be helpful, especially because otherwise this looks pretty similar to an iterator, but can be used in a slightly different and more efficient way w.r.t. `isFinished()/hasNext()`

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+	/**
+	 * @return true if end of input was reached, false otherwise.
+	 */
+	boolean isFinished();
+
+	/**
+	 * @return a future that is completed if there are more records available. If there are more
+	 * records available immediately, {@link #AVAILABLE} should be returned. Previously returned
+	 * not completed futures should become completed once there are more records available.
```
Comment (StefanRRichter) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java]: The interface does never talk about `records`, so I would rather go with `more input available`

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+ */
+
+package org.apache.flink.runtime.io;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Interface defining couple of essential methods for asynchronous and non blocking data polling.
+ */
+public interface AsyncDataInput<T> {
```
Comment (StefanRRichter) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java]: @Internal?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/BarrierBuffer.java [hunk 8]
```diff
@@ -152,37 +155,37 @@ public BarrierBuffer(InputGate inputGate, BufferBlocker bufferBlocker, long maxB
 		this.queuedBuffered = new ArrayDeque<BufferOrEventSequence>();
 	}
 
+	@Override
+	public CompletableFuture<?> isAvailable() {
+		if (currentBuffered == null) {
+			return inputGate.isAvailable();
```
Comment (StefanRRichter) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/BarrierBuffer.java]: I think this is not ok, `InputGate#AVAILABLE` is currently not the same as `AsyncDataInput#AVAILABLE`, which messes up the checks for referencial equality. I guess you want to replace the `InputGate#AVAILABLE` with the instance from the interface.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/BarrierBuffer.java [hunk 9]
```diff
@@ -207,11 +210,28 @@ else if (bufferOrEvent.getEvent().getClass() == CancelCheckpointMarker.class) {
...
 			}
 		}
 	}
 
+	private Optional<BufferOrEvent> handleEmptyBuffer() throws Exception {
+		if (!inputGate.isFinished()) {
+			return Optional.empty();
+		}
+
+		if (!endOfStream) {
```
Comment (StefanRRichter) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/BarrierBuffer.java]: Does not follow your advice to put the simple case first and even has to negate the predicate to do so. You could move the else branch first and out.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/BarrierBuffer.java [hunk 10]
```diff
@@ -207,11 +210,28 @@ else if (bufferOrEvent.getEvent().getClass() == CancelCheckpointMarker.class) {
...
+			// end of input stream. stream continues with the buffered data
```
Comment (StefanRRichter) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/BarrierBuffer.java]: This comment and interaction is not very straight forward IMO. Could we pull out the code that continues with the buffered data into a dedicated method that is a) used as part of `pollNext()` for that case and b) is used here instead of `pollNext()`? Because we already know we will end up in a sub-case of `pollNext`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierHandler.java [hunk 11]
```diff
@@ -19,34 +19,31 @@
...
-	 * @throws InterruptedException Thrown if the thread is interrupted while blocking during
-	 *                              waiting for the next BufferOrEvent to become available.
-	 * @throws Exception Thrown in case that a checkpoint fails that is started as the result of receiving
-	 *                   the last checkpoint barrier
+	 * Blocking version of {@link #pollNext()}.
 	 */
-	BufferOrEvent getNextNonBlocked() throws Exception;
+	@Deprecated
+	default BufferOrEvent getNextNonBlocked() throws Exception {
+		Optional<BufferOrEvent> bufferOrEvent = pollNext();
```
Comment (StefanRRichter) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierHandler.java]: I am a bit confused: this does not look like a blocking implementation (e.g. using `isAvailable()`) but rather like it could actually violate the conditions on the next line.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java [hunk 12]
```diff
@@ -206,27 +207,35 @@ public boolean processInput() throws Exception {
...
-					// Event received
-					final AbstractEvent event = bufferOrEvent.getEvent();
-					if (event.getClass() != EndOfPartitionEvent.class) {
-						throw new IOException("Unexpected event: " + event);
+			final Optional<BufferOrEvent> bufferOrEvent = barrierHandler.pollNext();
+			if (bufferOrEvent.isPresent()) {
+				processBufferOrEvent(bufferOrEvent.get());
+			} else {
+				if (!barrierHandler.isFinished()) {
+					barrierHandler.isAvailable().get();
```
Comment (StefanRRichter) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java]: This whole method is bit of a mess for my taste. Still wonder if we would not rewrite this part a little so that `processBufferOrEvent` is still called in the same iteration when availability is signaled and not after another roundtrip?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java (hunk 4 above)
Comment (StefanRRichter) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java]: Right, you cannot add a lot of information about that, reason is more that you can see declared exceptions in the docs if you add them. Just a visibility thing.

Comment (StefanRRichter) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java]: Both are established and ok, I just found that poll still at times makes me slightly think more which of the two it was :D

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+ *
+ * <p>For the most efficient usage, user of this class is suppose to call {@link #pollNext()}
+ * until it returns that no more elements are available. If that happens, he should check if
+ * input {@link #isFinished()}. If not, he should wait for {@link #isAvailable()}
+ * {@link CompletableFuture} to be completed. For example:
+ *
+ * <pre>
+ * {@code
+ *	while (!input.isFinished()) {
+ *		AsyncDataInput<Optional<T>> input = ...;
```
Comment (1u0) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java]: Move this assignment before the outer `while` loop?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+	 *         next returned element. {@code Optional.empty()} will be returned if there is no data
+	 *         to return or if {@link #isFinished()} returns true.
+	 *     </li>
+	 *     <li>
+	 *         If next element or {@code null} if there is no data to return or if
+	 *         {@link #isFinished()} returns true.
+	 *     </li>
+	 * </ol>
+	 */
+	T pollNext() throws Exception;
```
Comment (1u0) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java]: Instead of using `AsyncDataInput<Optional<T>>`, would it make sense to make this method explicitly return `Optional<T>`? This way, the contract in your java doc would be "enforced" by method signature explicitly.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+	 * @return
+	 * <ol>
+	 *     <li>
+	 *         If the {@link T} is an {@code Optional}: return {@code Optional.of(element)} -
+	 *         next returned element. {@code Optional.empty()} will be returned if there is no data
+	 *         to return or if {@link #isFinished()} returns true.
+	 *     </li>
+	 *     <li>
+	 *         If next element or {@code null} if there is no data to return or if
+	 *         {@link #isFinished()} returns true.
```
Comment (1u0) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java]: I personally find this documentation vague. Too many `if`s, hard to understand what is what.
Could you try to simplify this docs?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java (hunk 14 above)
Comment (1u0) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java]: The `Input` interface in #8476 PR uses `AsyncDataInput<StreamElement>` and uses `null` as sentinel (the interface doesn't use `Optional<StreamElement>` due to performance concerns).

We have discussed offline with @pnowojski and there were some alternative ideas how the interface of `AsyncDataInput` may be more explicit (so the contract  of `pollNext()` method is not overloaded by `Optional.empty()` or `null` return value as marker) and more cleaner (at least, from my point of view).

I'll just leave one of the ideas that I find fine (but not enforced).
The `AsyncDataInput<T>` could have been extended to have two `poll*` methods, with the same behavior regarding traversal of input, but with just different caller conventions. So it's up to the caller to decide which flavor they prefer: 
```java
interface AsyncDataInput<T> {
    ...
    Optional<T> pollNext();
    
    @Nullable
    default T pollNextOrNull() {
         return pollNext().orElse(null);
    }
}
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/BarrierBuffer.java [hunk 16]
```diff
@@ -112,6 +113,8 @@
 	/** Flag to indicate whether we have drawn all available input. */
 	private boolean endOfStream;
 
+	private boolean isFinished;
```
Comment (Aitozi) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/BarrierBuffer.java]: No comment for this field, I think this should be comment because it has a different behavior with `BufferTracker.java` about `finish`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java [hunk 17]
```diff
@@ -494,12 +486,12 @@ public void requestPartitions() throws IOException, InterruptedException {
 	// ------------------------------------------------------------------------
 
 	@Override
-	public Optional<BufferOrEvent> getNextBufferOrEvent() throws IOException, InterruptedException {
+	public Optional<BufferOrEvent> getNext() throws IOException, InterruptedException {
 		return getNextBufferOrEvent(true);
```
Comment (Aitozi) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java]: I notice here still have a choice for blocking or non-blocking, I check the code path, does this only used for the batch task?(just a question)

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java (hunk 12 above)
Comment (Aitozi) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java]: Does this means we will waste `the time of waiting for new data` delaying execute mail for the this roundtrip?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java (hunk 15 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java]: Do you have any suggestion how?

**Comment by StefanRRichter** [comment]
@pnowojski reminder that I think this is good to merge, you might have a look into other comments one more time.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java (hunk 14 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java]: I have split this into `AsyncDataInput` and `NullableAsyncDataInput`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java (hunk 17 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java]: I think yes. Blocking method looks like it is only used in `AbstractRecordReader` and/or in tests.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/BarrierBuffer.java [hunk 18]
```diff
@@ -152,37 +155,37 @@ public BarrierBuffer(InputGate inputGate, BufferBlocker bufferBlocker, long maxB
...
+		return AVAILABLE;
+	}
+
 	// ------------------------------------------------------------------------
 	//  Buffer and barrier handling
 	// ------------------------------------------------------------------------
 
 	@Override
-	public BufferOrEvent getNextNonBlocked() throws Exception {
+	public Optional<BufferOrEvent> pollNext() throws Exception {
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/BarrierBuffer.java]: To make it really hard for a user to forget about checking for an empty result. Alternate/equivalent solution would be to mark it `@Nullable` and enable compile time errors about mishandling `null` (but that's almost impossible to do in our code base with so much legacy code).

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java (hunk 12 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java]: @Aitozi at the moment yes. To make it work correctly with mailbox thread, this change of non blocking `BarrierHandler` has to be integrated with pending @StefanRRichter 's PR.

**Comment by pnowojski** [comment]
I've addressed the other comments and pushed the changes.

**Comment by Aitozi** [comment]
LGTM +1.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/AvailabilityListener.java [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+	 *	}
+	 * }
+	 * </pre>
+	 *
+	 * @return a future that is completed if there are more records available. If there are more
+	 * records available immediately, {@link #AVAILABLE} should be returned. Previously returned
+	 * not completed futures should become completed once there is more input available or if
+	 * the input {@link #isFinished()}.
+	 */
+	CompletableFuture<?> isAvailable();
```
Comment (1u0) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/AvailabilityListener.java]: Maybe you can simplify the contract of this interface, by having only one method that returns 3 possible states of an underlying data stream? In nutshell, implementations of this interface have three situations:
 * stream has finished;
 * stream has not finished, and there is an immediate data ready to be processed;
 * stream has not finished, but the caller needs to wait when data is ready.

One way, you can apply the same trick as with `CompletableFuture<?> AVAILABLE = CompletableFuture.completedFuture(null);`.
Basically, having pre-defined different constant dummy `CompletableFuture` that would be marker of `isFinished`:

```
public interface AvailabilityListener {
    CompletableFuture<?> FINISHED = CompletableFuture.completedFuture(null);
    CompletableFuture<?> AVAILABLE = CompletableFuture.completedFuture(null);

    CompletableFuture<?> getStatusFuture();
}
```

Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/AvailabilityListener.java]: At first I liked the idea, however:
1. I think it is actually more difficult for the class user. Having `isFinished()` method is pretty self explanatory, while you would have to read about `FINISHED` constant in the java doc. Yes, it solves the connection between `isFinished()` and `isAvailable()`, but the issue that I mentioned IMO kind of balances this out
2. It would be actually quite a bit harder to implement and could cause new potential issues/bugs, like someone switched `getStatusFuture()` to `FINISHED` and then accidentally back to `AVAILABLE`.

I think there is another solution, with exposing to the implementator some context class similar to the one used in mailbox, while keeping the outer interface named differently.

However same comment as I wrote to Stefan, I think this will have to be kept in sync with `SourceReader` and we should discuss it there.

Comment (StefanRRichter) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/AvailabilityListener.java]: There might be even another way that we can discuss here. IIRC one core problem in this PR with `CompletableFuture` is that it is a class and not an interface, so the implementation that is already completed cannot be more effcient. Did we ever consider returning a `CompletionStage` instead, which could be backed by a completable future in the cases that require waiting, and with some efficient completed implementations? All return an enum that indicates either `AVAILABLE` or `FINISHED`.

Comment (StefanRRichter) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/AvailabilityListener.java]: Something ala `CompletionStage<Status> inputStatus()`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/NullableAsyncDataInput.java [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+
+import javax.annotation.Nullable;
+
+/**
+ * The variant of {@link AsyncDataInput} that for performance reasons returns {@code null} from
+ * {@link #pollNextNullable()} instead returning {@code Optional.empty()} from
+ * {@link AsyncDataInput#pollNext()}.
+ */
+@Internal
+public interface NullableAsyncDataInput<T> extends AvailabilityListener {
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/NullableAsyncDataInput.java]: Yes, I agree method name is not clear in this regard, but I think the javadoc of the method clarifies this enough.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/AvailabilityListener.java (hunk 19 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/AvailabilityListener.java]: No, we haven't thought about this. This also might be a good idea. I'll merge this as it is now, and as I wrote before, let's revisit it later after depending PRs on this are merged with context of propose `SourceReader` in mind.
